File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/MenuFragment.java
package com.pennapps.labs.pennmobile;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;

import com.pennapps.labs.pennmobile.classes.DiningHall;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import butterknife.ButterKnife;

public class MenuFragment extends Fragment {

    TabAdapter pageAdapter;
    ViewPager pager;

    private DiningHall mDiningHall;
    private MainActivity mActivity;

    class TabAdapter extends FragmentStatePagerAdapter {

        ArrayList<HashMap<String, ArrayList<String>>> foods;  //for each meal: {name of station: arraylist of foods at the station}
        ArrayList<String> headers;
        String name;

        public TabAdapter(FragmentManager fm) {
            super(fm);
            foods = new ArrayList<HashMap<String, ArrayList<String>>>();
            headers = new ArrayList<>();
        }

        private void addTabs(DiningHall hall) {
            List<DiningHall.Menu> menus = hall.menus;
            name = hall.getName();
            headers.add("HOURS");
            foods.add(new HashMap<String, ArrayList<String>>());    //first menu is empty for dining hall info tab
            for (DiningHall.Menu menu: menus) {
                HashMap<String, ArrayList<String>> stations = new HashMap<String, ArrayList<String>>();
                headers.add(menu.name);
                for (DiningHall.DiningStation station : menu.stations) {
                    ArrayList<String> foods = new ArrayList<>();
                    StringBuilder foodItems = new StringBuilder();      //for design purposes
                    for (int i=0; i<station.items.size(); i++){
                        String txt = station.items.get(i).title;
                        foodItems.append(String.valueOf(txt.charAt(0)).toUpperCase());
                        foodItems.append(txt.substring(1, txt.length()));
                        if (i<station.items.size()-1) {
                            foodItems.append("\n");
                        }
                    }

                    foods.add(foodItems.toString());
                    stations.put(StringUtils.capitalize(station.name), foods);
                }
                foods.add(stations);
            }
        }


        @Override
        public Fragment getItem(int position) {
            Fragment myFragment;
            if (position==0){
                myFragment = new DiningInfoFragment();
                Bundle args = new Bundle();
                args.putParcelable("DiningHall",mDiningHall);
                args.putString("name", name);
                args.putString("type","info");
                myFragment.setArguments(args);
            }
            else{
                myFragment = new MenuTab();
                Bundle args = new Bundle();
                args.putString("type","menu");
                args.putString("name",name);
                args.putStringArrayList("stations", new ArrayList<String>(foods.get(position).keySet()));
                HashMap<String, ArrayList<String>> stations = foods.get(position);
                for (String station: stations.keySet()){
                    args.putStringArrayList(station, stations.get(station));
                }
                myFragment.setArguments(args);
            }
            return myFragment;
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return headers.get(position);
        }

        @Override
        public int getCount() {
            return foods.size();
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mDiningHall = getArguments().getParcelable("DiningHall");
        mActivity = (MainActivity) getActivity();
        setHasOptionsMenu(true);
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_menu, container, false);
        pageAdapter = new TabAdapter(getActivity().getSupportFragmentManager());
        pageAdapter.addTabs(mDiningHall);
        pager = (ViewPager) v.findViewById(R.id.menu_pager);
        pager.setAdapter(pageAdapter);

        v.setBackgroundColor(Color.WHITE);
        ButterKnife.bind(this, v);
        ((MainActivity) getActivity()).addTabs(pageAdapter, pager, true);
        return v;
    }
    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.dining, menu);
        super.onCreateOptionsMenu(menu, inflater);

    }

    @Override
    public void onResume() {
        super.onResume();
        mActivity.getActionBarToggle().setDrawerIndicatorEnabled(false);
        mActivity.getActionBarToggle().syncState();
        mActivity.setTitle(mDiningHall.getName());
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        getActivity().setTitle(R.string.dining);
        mActivity.removeTabs();
        ButterKnife.unbind(this);
    }
    @Override
    public void onDestroy() {
        super.onDestroy();
        mActivity.removeTabs();
    }
}

File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/MenuFragment.kt
package com.pennapps.labs.pennmobile

import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import com.pennapps.labs.pennmobile.classes.DiningHall
import org.apache.commons.lang3.StringUtils

class MenuFragment : Fragment() {
    private lateinit var mActivity: MainActivity
    private lateinit var toolBar: Toolbar

    private var mDiningHall: DiningHall? = null
    private var pageAdapter: PagerAdapter? = null

    inner class TabAdapter(
        fm: FragmentManager,
    ) : FragmentStatePagerAdapter(fm) {
        // for each meal: {name of station: arraylist of foods at the station}
        var foods: ArrayList<HashMap<String, ArrayList<String>>> = ArrayList()

        var headers: ArrayList<String> = ArrayList()
        var name: String? = null

        fun addTabs(hall: DiningHall?) {
            val menus = hall?.menus ?: ArrayList()
            name = hall?.name
            headers.add("HOURS")
            foods.add(HashMap()) // first menu is empty for dining hall info tab
            for (menu in menus) {
                val stations = HashMap<String, ArrayList<String>>()
                headers.add(menu.name)
                for (station in menu.stations) {
                    val foods = ArrayList<String>()
                    val foodItems = StringBuilder() // for design purposes
                    for (i in station.items.indices) {
                        val txt = station.items[i].title
                        foodItems.append(txt?.get(0)?.uppercaseChar())
                        foodItems.append(txt?.substring(1, txt.length))
                        if (i < station.items.size - 1) {
                            foodItems.append("\n")
                        }
                    }
                    foods.add(foodItems.toString())
                    stations[StringUtils.capitalize(station.name)] = foods
                }
                foods.add(stations)
            }
        }

        override fun getItem(position: Int): Fragment {
            val myFragment: Fragment
            if (position == 0) {
                myFragment = DiningInfoFragment()
                val args = Bundle()
                args.putParcelable("DiningHall", mDiningHall)
                args.putString(getString(R.string.menu_arg_name), name)
                myFragment.arguments = args
            } else {
                myFragment = MenuTab()
                val args = Bundle()
                args.putString(getString(R.string.menu_arg_name), name)
                args.putStringArrayList(getString(R.string.menu_arg_stations), ArrayList(foods[position].keys))
                val stations = foods[position]
                for (station in stations.keys) {
                    args.putStringArrayList(station, stations[station])
                }
                myFragment.arguments = args
            }
            return myFragment
        }

        override fun getPageTitle(position: Int): CharSequence = headers[position]

        override fun getCount(): Int = foods.size
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mDiningHall = arguments?.getParcelable("DiningHall")
        mActivity = activity as MainActivity
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        val v = inflater.inflate(R.layout.fragment_menu, container, false)
        pageAdapter = TabAdapter(mActivity.supportFragmentManager)
        (pageAdapter as TabAdapter).addTabs(mDiningHall)
        val pager: ViewPager = v.findViewById(R.id.menu_pager)
        pager.adapter = pageAdapter
        v.setBackgroundColor(Color.WHITE)
        mActivity.addTabs(pageAdapter as TabAdapter, pager, true)
        return v
    }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        super.onViewCreated(view, savedInstanceState)
        toolBar = mActivity.findViewById(R.id.toolbar)
        toolBar.visibility = View.VISIBLE
        mActivity.hideBottomBar()
    }

    override fun onCreateOptionsMenu(
        menu: Menu,
        inflater: MenuInflater,
    ) {
        inflater.inflate(R.menu.dining, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            mActivity.onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        mActivity.title = mDiningHall?.name
        if (mActivity.supportActionBar != null) {
            mActivity.supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onDestroyView() {
        if (view != null) {
            val pager: ViewPager = requireView().findViewById(R.id.menu_pager)
            pager.adapter = null
        }

        super.onDestroyView()
        mActivity.removeTabs()
        if (mActivity.supportActionBar != null) {
            mActivity.supportActionBar?.setDisplayHomeAsUpEnabled(false)
        }
        mActivity.supportActionBar?.hide()
    }

    override fun onDestroy() {
        super.onDestroy()
        mActivity.removeTabs()
    }
}
